### PR TITLE
cxx-qt-gen: allow for unused unsafe blocks and resolve no-exception issues

### PR DESCRIPTION
**Pull request by ahayzen-kdab**

### CODE REVIEW

File: crates/cxx-qt/src/lib.rs [hunk 1]
```diff
@@ -30,6 +31,21 @@ pub use threading::CxxQtThread;
...
+/// Errors that can occur from CXX-Qt
+#[derive(Error, Debug)]
+#[non_exhaustive]
+pub enum CxxQtError {
+    /// Threading failed as the object has been destroyed
+    #[error("Cannot queue function pointer as object has been destroyed")]
+    ThreadingQueueObjectDestroyed,
+    /// Threading failed calling invokeMethod on the object
+    #[error("Cannot queue function pointer as invokeMethod on object failed")]
+    ThreadingQueueInvokeMethod,
```
Comment (ahayzen-kdab) [inline crates/cxx-qt/src/lib.rs:43]: InvokeMethodFailed

File: crates/cxx-qt/src/lib.rs [hunk 2]
```diff
@@ -30,6 +31,21 @@ pub use threading::CxxQtThread;
 #[doc(hidden)]
 pub use static_assertions;
 
+/// Errors that can occur from CXX-Qt
+#[derive(Error, Debug)]
+#[non_exhaustive]
+pub enum CxxQtError {
```
Comment (ahayzen-kdab) [inline crates/cxx-qt/src/lib.rs:37]: consider cxx enum

Comment (ahayzen-kdab) [reply crates/cxx-qt/src/lib.rs]: This becomes complex and starts needing to expose types publicly and avoiding collisions when there are multiple QObjects in generated bridges etc. So just leave as an integer for now.

File: crates/cxx-qt/include/thread.h [hunk 3]
```diff
@@ -52,15 +54,13 @@ class CxxQtThread final
...
+  ::std::uint8_t queue(::rust::Fn<void(T& self, ::rust::Box<A> arg)> func,
+                       ::rust::Box<A> arg) const
   {
     // Ensure that we can read the pointer and it's not being written to
     const auto guard = ::std::shared_lock(m_obj->mutex);
     if (!m_obj->ptr) {
-      throw ::std::runtime_error(
-        "Cannot queue function pointer as object has been destroyed");
-      return;
+      return 1;
```
Comment (LeonMatthesKDAB) [inline crates/cxx-qt/include/thread.h:63]: nit: Magic number.
We should probably create constants for these on the C++ side to mirror the Rust enum.

e.g.
```cpp
const static ::std::uint8_t ObjectDestroyed = 1;
const static ::std::uint8_t InvokeMethodFailed = 1;
```

**Review by LeonMatthesKDAB — COMMENTED** [review]
Pretty much looks good, just a minor nit left.

File: book/src/concepts/wasm-builds.md [hunk 4]
```diff
@@ -190,3 +190,12 @@ error[E0433]: failed to resolve: use of undeclared type `IovLen`
...
 There is discussion around supporting wasm in the GitHub repository for `async-std`, and the progress is being tracked [here](https://github.com/async-rs/async-std/issues/220).
+
+### Cannot use throw / try catch with exceptions disabled
+
+CXX and CXX-Qt can generate code that attempts to use exceptions but these are disabled
+for most WASM platforms.
+
+The cmake-qt-cmake builder will attempt to set a define of `RUST_CXX_NO_EXCEPTIONS` which should
+causes the code not to be compiled. But if this doesn't work try adding `-DRUST_CXX_NO_EXCEPTIONS`
+to a `CXXFLAGS` enviornment variable while building.
```
[the remark this answers is not in the document]
Comment (LeonMatthesKDAB) [reply book/src/concepts/wasm-builds.md]: Forgot to post this on the last change, irrelevant now :)

File: crates/cxx-qt/include/thread.h (hunk 3 above)
Comment (ahayzen-kdab) [reply crates/cxx-qt/include/thread.h]: done

**Review by LeonMatthesKDAB — APPROVED** [review]

**Pull request merged**